### PR TITLE
relocate the invalid file link

### PR DESCRIPTION
**Pull request by live1206**
https://dev.azure.com/ceapex/Engineering/_workitems/edit/104748
https://dev.azure.com/ceapex/Engineering/_workitems/edit/105838

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4892)

### CODE REVIEW

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 1]
```diff
@@ -174,7 +174,10 @@ internal class DependencyResolver
                 file = TryResolveResourceFromHistory(_gitCommitProvider, declaringFile.Docset, pathToDocset, _templateEngine);
                 if (file is null)
                 {
-                    return (error, href, fragment, linkType, null, false);
+                    // the relative path can just be href when the migration from v2 to v3 is done
+                    var extensionIndex = pathToDocset.IndexOf('.');
+                    var pathToDocsetWithoutExtension = extensionIndex != -1 ? pathToDocset.Substring(0, extensionIndex) : pathToDocset;
```
Comment (OsmondJiang) [inline src/docfx/build/dependency/DependencyResolver.cs]: Can we reuse some methods in `Document` like `FilePathToSitePath`?

File: docs/specs/markdown.yml [hunk 2]
```diff
@@ -536,4 +536,21 @@ inputs:
     [some text](#)
 outputs:
   docs/a.json:
+---
+# invalid file link should be outputted as relative path
+inputs:
+  docfx.yml:
+  docs/1/a.md: |
+    [non-existed file](~/c.md)
+    [non-existed file](~/2/d)
```
Comment (OsmondJiang) [inline docs/specs/markdown.yml]: add a case for relative path not found?

**Comment by yufeih** [comment]
#4450

File: src/docfx/build/dependency/DependencyResolver.cs [hunk 3]
```diff
@@ -174,7 +174,16 @@ internal class DependencyResolver
...
-                    return (error, href, fragment, linkType, null, false);
+                    if (linkType != LinkType.RelativePath)
+                    {
+                        return (error, href, fragment, linkType, null, false);
+                    }
+
+                    // in v2, the non-reoslved link will be relocated to a relative path, v3 tries to avoid the behavior change here
+                    // the relocation of relative path can be removed when migration from v2 to v3 is done
+                    var relativePath = PathUtility.GetRelativePathToFile(declaringFile.FilePath, pathToDocset);
+                    var relativeUrl = Document.PathToRelativeUrl(relativePath, ContentType.Page, null, declaringFile.Docset.Config.Output.Json, true);
```
Comment (OsmondJiang) [inline src/docfx/build/dependency/DependencyResolver.cs]: Document.GetContentType

Comment (OsmondJiang) [reply src/docfx/build/dependency/DependencyResolver.cs]: Add spec test for resource not found

File: docs/specs/markdown.yml [hunk 4]
```diff
@@ -536,4 +536,21 @@ inputs:
     [some text](#)
 outputs:
   docs/a.json:
+---
+# invalid file link should be outputted as relative path
+inputs:
+  docfx.yml:
+  docs/1/a.md: |
```
Comment (928PJY) [inline docs/specs/markdown.yml]: what if `index.md`/ `../index.md` not found

Comment (live1206) [reply docs/specs/markdown.yml]: it has been handled, test case added for verification

**Pull request merged**